File: backend/app/names.py
"""How a person is named on screen, decided in exactly one place.

A user has a `username`, which is their unique login and never changes shape,
and an optional `display_name`, which is what they would rather be called.
Everywhere a person is named the second wins when it is set — but it is
nullable, it is not unique, and two people really can both be "Walt", so the
username stays the thing that disambiguates and the thing you sign in with.

The rule lives here rather than being spelled `display_name or username` at
each call site, because a call site that forgets it does not fail: it silently
shows the login name to somebody who set a display name, and the bug is
invisible until a user complains. One import is easier to grep for than a
scattered idiom, and the frontend keeps the same single resolver in
`src/names.ts` for the same reason.
"""
from __future__ import annotations

from .models import User
from .schemas import UserSummary

#: Longest display name we store. Wider than a username (64) because a real
#: name with a team in brackets is a reasonable thing to want.
MAX_DISPLAY_NAME = 128


def display_name(user: User | None) -> str:
    """What to call this person. Never empty, never None."""
    if user is None:
        return "someone"
    chosen = (user.display_name or "").strip()
    return chosen or user.username


def clean_display_name(raw: str | None) -> str | None:
    """Normalise an incoming display name, or None to clear it.

    Whitespace-only is stored as NULL rather than as a blank string: a blank
    would satisfy `display_name or username` in every naive check and render as
    an anonymous gap on screen.
    """
    if raw is None:
        return None
    trimmed = raw.strip()
    return trimmed[:MAX_DISPLAY_NAME] or None


def summarise(user: User) -> UserSummary:
    """The shape every screen that names other people reads.

    Carries both names: the resolver on the client needs the fallback, and the
    username is what tells two people called "Walt" apart.
    """
    return UserSummary(id=user.id, username=user.username, display_name=user.display_name)

File: frontend/src/names.ts
/**
 * How a person is named on screen, decided in exactly one place.
 *
 * A user has a `username` — their unique login — and an optional
 * `display_name`, which is what they would rather be called. The display name
 * wins wherever a person is named, but it is nullable and it is deliberately
 * *not* unique: two people called "Walt" is a real situation, and the username
 * is what tells them apart.
 *
 * This exists as one function rather than `u.display_name || u.username`
 * scattered through the views because the scattered form does not fail when it
 * is missed — it quietly shows a login name to somebody who chose another name,
 * in one screen out of eight, and nobody notices until they complain. The
 * server keeps the matching single resolver in `app/names.py`.
 */

/** The least a thing must have to be nameable. `UserSummary` and `User` both do. */
export type Named = { username: string; display_name?: string | null };

/** What to call this person. Never empty. */
export function displayName(person: Named | null | undefined): string {
  if (!person) return "someone";
  const chosen = (person.display_name ?? "").trim();
  return chosen || person.username;
}

/**
 * The name, plus the username in brackets when they differ.
 *
 * For the places where picking the wrong person has a consequence — granting
 * access to a stored credential, handing a session over — because display
 * names are not unique and "Walt" alone is not enough to choose between two
 * of them.
 */
export function fullName(person: Named | null | undefined): string {
  if (!person) return "someone";
  const shown = displayName(person);
  return shown === person.username ? shown : `${shown} (${person.username})`;
}

/** Look an id up in a directory and name them. `unknown` covers a stale id. */
export function nameById<T extends Named & { id: number }>(
  people: T[],
  id: number | null | undefined,
  unknown = "someone",
): string {
  if (id === null || id === undefined) return unknown;
  const found = people.find((p) => p.id === id);
  return found ? displayName(found) : unknown;
}
